docs/guides/building-forms.md

```diff
@@ -0,0 +1,387 @@
+
+Next.js offers a file-based system for routing that's built on the [concept of pages](/docs/basic-features/pages). Any file inside the folder `pages/api` is mapped to `/api/*` and will be treated as an API endpoint instead of a page. This API endpoint is going to be server-side only.
+
+Go to `pages/api`, create a file called `form.js` and paste this code written in Node.js:
+
+```js
+export default function handler(req, res) {
+  // Get data submitted in request's body.
+  const body = req.body
+
+  // Optional logging to see the responses
+  // in the command line where next.js app is running.
+  console.log('body: ', body)
+
+  // Both of these are required.
+  if (!body.first || !body.last) {
+    return res.json({ data: 'First or last name not found' })
+  }
+
+  // Found the name.
+  res.json({ data: `${body.first} ${body.last}` })
+}
+```
+
+This serverless form `handler` function will receive the request `req` from the client (i.e. submitted form data). And in return, it'll send a response `res` as JSON that will have both the first and the last name. You can access this API endpoint at `http://localhost:3000/api/form` or replace the localhost URL with an actual Vercel deployment when you deploy.
+
+> Moreover, you can also attach this API to a database like MongoDB or Google Sheets. This way, your submitted form data will be securely stored for later use. For this guide, no database is used. Instead, the same data is returned to the user to demo how it's done.
+
+### Form Submission without JavaScript
+
+You can now use `/api/form` relative endpoint inside the `action` attribute of the form. You are sending form data to the server (serverless API at Vercel) when the form is submitted via `POST` HTTP method (which is used to send data).
+
+```html
+<form action="/api/form" method="post">
+  <label for="first">First name:</label>
+  <input type="text" id="first" name="first" />
+  <label for="last">Last name:</label>
+  <input type="text" id="last" name="last" />
+  <button type="submit">Submit</button>
+</form>
+```
+
+If you submit this form, it will submit the data to the forms API endpoint `/api/form`. The server then responds, generally handling the data and loading the URL defined by the action attribute, causing a new page load. So in this case you'll be redirected to `http://localhost:3000/api/form` with the following response from the server.
+
+![form-no-js](https://assets.vercel.com/image/upload/c_scale,w_675/v1643009087/nextjs/guides/building-forms/form-no-js.jpg)
+
+## Part 3: Forms in Next.js
+
+You have created a server with Vercel via Serverless Functions. Now it's time to configure the client (the form itself) inside Next.js using React. The first step will be extending your knowledge of HTML forms and converting it into React (using [JSX](https://reactjs.org/docs/introducing-jsx.html)).
```

Comment:
I find this confusing as the above section `Part 2` says that we are going to create a serverless function using Vercel. In actual fact we have created an api route using next.js. Once deployed to vercel, it would use vercel infra, but currently its just an api route. I wonder if we need to re-work, or re-name the above section? 🤔

Reply:
Ok, now I understand what's making this confusing. My line of thought was:

- To keep Part 1 purely HTML-specific. Everything related to the basic form element, including the validation (both with and without JS), should go here. 
- Part 2 discusses the details about React/Next.js forms, which again re-aligns validation steps accordingly. E.g., how does form validation looks like with React. And then an explanation about setting up serverless env as well. 

Let's remove the [CLIENT] and [SERVER] parts from the headings. I hope that will make things clear. 🤔